contrib/zenith/libpagestore.c

```diff
@@ -50,10 +52,79 @@ page_server_api api = {
 static void
 zenith_connect()
 {
-	char	   *query;
-	int			ret;
+	char			 *query;
+	int				  ret;
+	char			 *auth_token;
+	char			 *err = NULL;
+	PQconninfoOption *conn_options;
+	PQconninfoOption *conn_option;
+	int 			 noptions = 0;
+
+    // this is heavily inspired by psql/command.c::do_connect
+	conn_options = PQconninfoParse(
+		page_server_connstring,
+	 	&err
+	);
+
+	if (conn_options == NULL) {
+		/* The error string is malloc'd, so we must free it explicitly */
+		char	   *errcopy = err ? pstrdup(err) : "out of memory";
+		PQfreemem(err);
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("invalid connection string syntax: %s", errcopy)));
+	}
+
+	auth_token = getenv(ZENITH_AUTH_TOKEN_ENV);
+	if (!auth_token) {
+		zenith_log(LOG, "No auth_token in env");
+	} else {
+		zenith_log(LOG, "auth_token passed via env");
+		for (conn_option = conn_options; conn_option->keyword != NULL; conn_option++)
+		{
+			if (strcmp(conn_option->keyword, "password") == 0)
+			{
+				if (conn_option->val != NULL && conn_option->val[0] != '\0')
+				{
+					// check for placeholder value
+					if (strcmp(conn_option->val, "ztoken") == 0)
+					{
+						// inspired by PQconninfoFree and conninfo_storeval
+						// so just free the old one and replace with freshly malloc'ed one
+						free(conn_option->val);
+						conn_option->val = strdup(auth_token);
+					}
+					break;
+				}
+			}
+		}
+	}
+
+	// walk the NULL terminated array to determine how many options we have
+	for (conn_option = conn_options; conn_option->keyword != NULL; conn_option++)
+	{
+		noptions++;
+	}
+
+	// copy values from PQconninfoOption to key/value arrays because PQconnectdbParams accepts options this way
+	const char **keywords = malloc((noptions + 1) * sizeof(*keywords));
+	const char **values = malloc((noptions + 1) * sizeof(*values));
+	int			 i = 0;
+	
+	for (conn_option = conn_options; conn_option->keyword != NULL; conn_option++, i++)
+	{
+		keywords[i] = conn_option->keyword;
+		values[i] = conn_option->val;
+	}
+	// add array terminator
+	keywords[i] = NULL;
+	values[i] = NULL;
+
+	pageserver_conn = PQconnectdbParams(keywords, values, false);
+	free(keywords);
+	free(values);
```

Comment:
It would be also good to add comment somewhere around `auth_token = getenv(ZENITH_AUTH_TOKEN_ENV)` that states that follow up code is to add ZENITH_AUTH_TOKEN to connection options